### PR TITLE
TypeScript definitions

### PR DESCRIPTION
**Pull request by emlynmac**
This PR adds TS definitions in one export, which is exported via types in the package.json.
It is meant to improve on #142 and ratify #146 
I've tested this via including from my fork to package.json:

```
"dependencies": {
  ...,
  "html-to-react": "emlynmac/html-to-react#ts-defs"
}
```

### CODE REVIEW

**Comment by aknuds1** [comment]
Could you please explain how this PR relates to #146, is it supposed to replace it? 

If it should replace my PR, should it include my rewrite of the tests to TypeScript?

**Comment by emlynmac** [comment]
> Could you please explain how this PR relates to #146, is it supposed to replace it?
> 
> If it should replace my PR, should it include my rewrite of the tests to TypeScript?

It is intended to replace the TS definitions. The  autogenerated ones were not ideal. 

Yes, it should include the updated test file too.

**Comment by aknuds1** [comment]
Can you update this PR to include the TypeScript tests from the other PR? I.e., make it equivalent to the other PR sans the auto-generated TS definitions.

**Comment by emlynmac** [comment]
> Can you update this PR to include the TypeScript tests from the other PR? I.e., make it equivalent to the other PR sans the auto-generated TS definitions.

Maybe you'd prefer to take this one file and bring it into your PR as you've started the process already and your PR is more complete?

**Comment by aknuds1** [comment]
@emlynmac how do I make your `html-to-react` module work with the tests? After rebasing your PR on mine, so your type definitions are used and my tests are used, I get the following TypeScript compilation error on trying to run the tests:

``` console
$ npm test                    

> html-to-react@1.5.1 test
> ./node_modules/.bin/eslint . && ./node_modules/.bin/mocha


TSError: ⨯ Unable to compile TypeScript:
test/html-to-react-tests.ts:5:22 - error TS2306: File '/home/arve/Projects/html-to-react/index.d.ts' is not a module.

5 import {Parser} from '..';
                       ~~~~
test/html-to-react-tests.ts:6:38 - error TS2306: File '/home/arve/Projects/html-to-react/index.d.ts' is not a module.

6 import {ProcessNodeDefinitions} from '..';
```

**Comment by emlynmac** [comment]
> @emlynmac how do I make your `html-to-react` module work with the tests? After rebasing your PR on mine, so your type definitions are used and my tests are used, I get the following TypeScript compilation error on trying to run the tests:
> 
> ```
> $ npm test                    
> 
> > html-to-react@1.5.1 test
> > ./node_modules/.bin/eslint . && ./node_modules/.bin/mocha
> 
> 
> TSError: ⨯ Unable to compile TypeScript:
> test/html-to-react-tests.ts:5:22 - error TS2306: File '/home/arve/Projects/html-to-react/index.d.ts' is not a module.
> 
> 5 import {Parser} from '..';
>                        ~~~~
> test/html-to-react-tests.ts:6:38 - error TS2306: File '/home/arve/Projects/html-to-react/index.d.ts' is not a module.
> 
> 6 import {ProcessNodeDefinitions} from '..';
> ```

I will have to have a look into this a bit more when I get some time. I've not used TypeScript tests in a JS module, and it looks like at first glance that the index.d.ts and index.js are clashing.
For testing TS defs in JS libraries, there is this library: https://github.com/ai/check-dts

**Comment by emlynmac** [comment]
> @emlynmac how do I make your `html-to-react` module work with the tests? After rebasing your PR on mine, so your type definitions are used and my tests are used, I get the following TypeScript compilation error on trying to run the tests:
> 
> ```
> $ npm test                    
> 
> > html-to-react@1.5.1 test
> > ./node_modules/.bin/eslint . && ./node_modules/.bin/mocha
> 
> 
> TSError: ⨯ Unable to compile TypeScript:
> test/html-to-react-tests.ts:5:22 - error TS2306: File '/home/arve/Projects/html-to-react/index.d.ts' is not a module.
> 
> 5 import {Parser} from '..';
>                        ~~~~
> test/html-to-react-tests.ts:6:38 - error TS2306: File '/home/arve/Projects/html-to-react/index.d.ts' is not a module.
> 
> 6 import {ProcessNodeDefinitions} from '..';
> ```

I've just pushed up some changes;
In VS Code, I see that the types are picked up for the tests but not from the command line.
I''m also seeing lots of issues with typing due to the parser class coming back with either an object or array of objects.

**Comment by aknuds1** [comment]
I've tried executing the tests in your branch, and they all fail with type errors, e.g.

```console
test/html-to-react-tests.ts:16:61 - error TS2345: Argument of type 'Element | Element[]' is not assignable to parameter of type 'ReactElement<any, string | JSXElementConstructor<any>>'.
  Type 'Element[]' is missing the following properties from type 'ReactElement<any, string | JSXElementConstructor<any>>': type, props, key

16       const reactHtml = ReactDOMServer.renderToStaticMarkup(reactComponent);
                                                               ~~~~~~~~~~~~~~
```

I guess the type definitions need some work.

**Comment by emlynmac** [comment]
> I've tried executing the tests in your branch, and they all fail with type errors, e.g.
> 
> ```
> test/html-to-react-tests.ts:16:61 - error TS2345: Argument of type 'Element | Element[]' is not assignable to parameter of type 'ReactElement<any, string | JSXElementConstructor<any>>'.
>   Type 'Element[]' is missing the following properties from type 'ReactElement<any, string | JSXElementConstructor<any>>': type, props, key
> 
> 16       const reactHtml = ReactDOMServer.renderToStaticMarkup(reactComponent);
>                                                                ~~~~~~~~~~~~~~
> ```
> 
> I guess the type definitions need some work.

The type definitions are now following what the original JS code is doing more closely. This has broken the tests that I merged from your original PR.

The issue you have highlighted is due to the return value here: 
https://github.com/aknuds1/html-to-react/blob/master/lib/parser.js#L62

This line will return either an array or just the first one if the length is only one. This means the return type will be `JSX.Element | JSX.Element[]` as it can be either.

The workaround is to unpack the array if it is, or simply pass along the object:
```
      const reactComponent = parser.parse(htmlInput);
      const reactHtml = Array.isArray(reactComponent) ?
         ReactDOMServer.renderToStaticMarkup(reactComponent[0])
      : ReactDOMServer.renderToStaticMarkup(reactComponent);
```

The second issue with the props not being found is due to the same issue of potentially passing back either an array or an object. `props` exists on the `Element` type, but not on the array type, which gets the compiler upset.
Again, the fix is to unpack the potential array return type.

The last issue was a missed type.

It might be worth rethinking the return type of the parser to always hand back an array of elements, for clarity and API ergonomics.

File: .mocharc.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "extension": ["ts"],
+  "spec": "test/**/*-tests.ts",
+  "require": "ts-node/register"
+}
```
Comment (aknuds1) [inline .mocharc.json:5]: Could you fix this file to end with a newline, like in my TS definitions PR?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Change Log
 
 ## Master/Unreleased
-
+Add TypeScript type definitions [#146](https://github.com/emlynmac/html-to-react/pull/146) ([xnevz](https://github.com/xnevz), [aknuds1](https://github.com/aknuds1),[emlynmac](https://github.com/emlynmac))
```
Comment (aknuds1) [inline CHANGELOG.md:4]: Please add a blank line before and after the entry, and change github.com/emlynmac into github.com/aknuds1.

File: package.json [hunk 3]
```diff
@@ -41,13 +41,20 @@
     "react": "^18.0"
   },
   "devDependencies": {
+    "@types/mocha": "^10.0.1",
+    "@types/react": "^18.0.33",
+    "@types/react-dom": "^18.0.11",
+    "check-dts": "^0.7.2",
```
Comment (aknuds1) [inline package.json:47]: What's the purpose of this new dependency? I don't see it in use.

File: test/boolattrs.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+"use strict";
```
Comment (aknuds1) [inline test/boolattrs.ts:1]: What's the need of `use strict` with TypeScript? Isn't strict mode already enabled in tsconfig.json?

File: test/boolattrs.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+// Boolean HTML attributes, in the case React expects.
+const booleanAttrs = [
+  "allowFullScreen",
```
Comment (aknuds1) [inline test/boolattrs.ts:5]: Why are you diverging from the project's style of using single quotes? Could you revert to single quotes instead, so we avoid this intrusive change?

File: test/html-to-react-tests.ts [hunk 6]
```diff
@@ -0,0 +1,681 @@
+import { equal, strictEqual, deepStrictEqual, ok } from "assert";
```
Comment (aknuds1) [inline test/html-to-react-tests.ts:1]: Why are you diverging from the project's style of using single quotes? Could you revert to single quotes instead, so we avoid this intrusive change?

**Review by aknuds1 — COMMENTED** [review]
Please see comments.

**Review by aknuds1 — COMMENTED** [review]
Please see comments.

**Review by aknuds1 — COMMENTED** [review]
Please see comments.

**Review by aknuds1 — COMMENTED** [review]
Please see comments.

**Comment by aknuds1** [comment]
Wow, weird that GitHub submitted my review four(!) times. Also, tests still won't run as you can see from CI (same thing happens locally for me).

File: CHANGELOG.md (hunk 2 above)
Comment (emlynmac) [reply CHANGELOG.md:4]: Feel free to update any stylistic changes as needed.

File: .mocharc.json (hunk 1 above)
Comment (aknuds1) [reply .mocharc.json:5]: I'm sorry, but that's not my job.

File: CHANGELOG.md (hunk 2 above)
Comment (aknuds1) [reply CHANGELOG.md:4]: I'm sorry, but that's not my job.

**Comment by aknuds1** [comment]
> To be clear; they're not 'stricter' they're just more accurate. The JS APIs are returning both single object and array types, so if you don't include this in your type definitions, you're basically breaking the TS definitions.

I don't agree with your view. As I see it, your `parse` type definition specifically is stricter in not using the loose `any` return type. This is better than my approach (just going with auto-generated type definitions, and `any` for the return type) in theory, but your PR just doesn't compile, so it simply can't be merged.

> If you decide to merge in your PR and TS definitions instead, please make ensure they are describing the APIs correctly.

I've merged my PR. I don't agree that the type definitions in there don't describe the API correctly; they employ looser, auto-generated, typing (see my first paragraph), but that doesn't make them _incorrect_. As the TypeScript compiler compiles the code with those definitions, I fail to see how they could be incorrect. Also, bear in mind that the type definitions were automatically generated, did the compiler generate incorrect definitions?

If you want to improve the type definitions, you're welcome to, but they have to at the very least compile.

**Comment by aknuds1** [comment]
Closing due to inactivity.

**Pull request closed without merging**